src/scheduler/Scheduler.cpp

```diff
@@ -535,18 +561,54 @@ std::vector<faabric::Message> Scheduler::getRecordedMessagesLocal()
     return recordedMessagesLocal;
 }
 
+std::string getClientKey(const std::string& otherHost)
+{
+    // Note, our keys here have to include the tid as the clients can only be
+    // used within the same thread.
+    std::thread::id tid = std::this_thread::get_id();
+    std::stringstream ss;
+    ss << otherHost << "_" << tid;
+    std::string key = ss.str();
+    return key;
+}
+
 FunctionCallClient& Scheduler::getFunctionCallClient(
   const std::string& otherHost)
 {
-    auto it = functionCallClients.find(otherHost);
-    if (it == functionCallClients.end()) {
-        auto _it = functionCallClients.try_emplace(otherHost, otherHost);
-        if (!_it.second) {
-            throw std::runtime_error("Error inserting function call client");
+    std::string key = getClientKey(otherHost);
+    if (functionCallClients.find(key) == functionCallClients.end()) {
+        faabric::util::FullLock lock(functionCallClientsMx);
+
+        if (functionCallClients.find(key) == functionCallClients.end()) {
+            logger->debug(
+              "Adding new function call client for {} ({})", otherHost, key);
+            functionCallClients.emplace(key, otherHost);
+        }
+    }
+
+    {
+        faabric::util::SharedLock lock(functionCallClientsMx);
+        return functionCallClients.at(key);
+    }
+}
+
+SnapshotClient& Scheduler::getSnapshotClient(const std::string& otherHost)
+{
+    std::string key = getClientKey(otherHost);
+    if (snapshotClients.find(key) == snapshotClients.end()) {
+        faabric::util::FullLock lock(snapshotClientsMx);
+
+        if (snapshotClients.find(key) == snapshotClients.end()) {
+            logger->debug(
+              "Adding new snapshot client for {} ({})", otherHost, key);
+            snapshotClients.emplace(key, otherHost);
         }
-        it = _it.first;
     }
-    return it->second;
+
+    {
+        faabric::util::SharedLock lock(snapshotClientsMx);
+        return snapshotClients.at(key);
+    }
```

Comment:
These were the changes to `getSnapshotClient` and `getFunctionCallClient` that address the issues of cross-thread clients and race conditions.